File: app/src/androidTest/java/it/cosenonjaviste/ui/CnjDaggerRule.java
package it.cosenonjaviste.ui;

import android.os.AsyncTask;
import android.support.test.InstrumentationRegistry;

import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.Statement;

import io.reactivex.Scheduler;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import it.cosenonjaviste.daggermock.DaggerMockRule;
import it.cosenonjaviste.model.TwitterService;
import it.cosenonjaviste.model.WordPressService;

public class CnjDaggerRule extends DaggerMockRule<ApplicationComponent> {
    public CnjDaggerRule() {
        super(ApplicationComponent.class, new AppModule(getApp()));
        providesMock(WordPressService.class, TwitterService.class);
        set(component -> getApp().setComponent(component));
    }

    public static CoseNonJavisteApp getApp() {
        return (CoseNonJavisteApp) InstrumentationRegistry.getInstrumentation().getTargetContext().getApplicationContext();
    }

    @Override public Statement apply(Statement base, FrameworkMethod method, Object target) {
        Statement superStatement = super.apply(base, method, target);
        Scheduler asyncTaskScheduler =
                Schedulers.from(AsyncTask.THREAD_POOL_EXECUTOR);
        return new Statement() {
            @Override public void evaluate() throws Throwable {
                RxJavaPlugins.setIoSchedulerHandler(
                        scheduler -> asyncTaskScheduler);
                RxJavaPlugins.setComputationSchedulerHandler(
                        scheduler -> asyncTaskScheduler);
                RxJavaPlugins.setNewThreadSchedulerHandler(
                        scheduler -> asyncTaskScheduler);

                try {
                    superStatement.evaluate();
                } finally {
                    RxJavaPlugins.reset();
                }
            }
        };
    }
}

File: app/src/androidTest/java/it/cosenonjaviste/ui/CnjDaggerRule.kt
package it.cosenonjaviste.ui

import android.support.test.InstrumentationRegistry
import it.cosenonjaviste.androidtest.base.EspressoExecutor
import it.cosenonjaviste.daggermock.DaggerMockRule
import it.cosenonjaviste.model.TwitterService
import it.cosenonjaviste.model.WordPressService
import org.junit.runners.model.FrameworkMethod
import org.junit.runners.model.Statement
import rx.Scheduler
import rx.android.plugins.RxAndroidPlugins
import rx.android.plugins.RxAndroidSchedulersHook
import rx.plugins.RxJavaHooks
import rx.schedulers.Schedulers

class CnjDaggerRule : DaggerMockRule<ApplicationComponent>(ApplicationComponent::class.java, AppModule(app)) {
    init {
        providesMock(WordPressService::class.java, TwitterService::class.java)
        set { component -> app.component = component }
    }

    override fun apply(base: Statement, method: FrameworkMethod?, target: Any): Statement {
        val superStatement = super.apply(base, method, target)
        return object : Statement() {
            @Throws(Throwable::class)
            override fun evaluate() {
                RxJavaHooks.setOnIOScheduler { scheduler -> Schedulers.immediate() }
                RxJavaHooks.setOnComputationScheduler { scheduler -> Schedulers.immediate() }
                RxJavaHooks.setOnNewThreadScheduler { scheduler -> Schedulers.immediate() }
                RxAndroidPlugins.getInstance().registerSchedulersHook(object : RxAndroidSchedulersHook() {
                    override fun getMainThreadScheduler(): Scheduler {
                        return Schedulers.from(EspressoExecutor)
                    }
                })
                try {
                    superStatement.evaluate()
                } finally {
                    RxJavaHooks.reset()
                    RxAndroidPlugins.getInstance().reset()
                }
            }
        }
    }

    companion object {

        val app: CoseNonJavisteApp
            get() = InstrumentationRegistry.getInstrumentation().targetContext.applicationContext as CoseNonJavisteApp
    }
}
